libyul/optimiser/ZeroByReturndatasizeReplacer.cpp

```diff
@@ -0,0 +1,164 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/optimiser/ZeroByReturndatasizeReplacer.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/SideEffects.h>
+#include <libyul/Utilities.h>
+#include <libyul/AST.h>
+
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/reverse.hpp>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace
+{
+bool functionContainsCall(Dialect const& _dialect, map<YulString, SideEffects> const& _functionSideEffects, FunctionCall const& _funCall) {
+	if (BuiltinFunction const* builtin = _dialect.builtin(_funCall.functionName.name))
+		if (!builtin->sideEffects.keepsReturndatasize)
+			return true;
+	if (auto const* sideEffects = util::valueOrNullptr(_functionSideEffects, _funCall.functionName.name))
+		if (!sideEffects->keepsReturndatasize)
+			return true;
+	return false;
+}
+
+class CallTracer: public ASTModifier
+{
+public:
+	CallTracer(
+		Dialect const& _dialect,
+		map<YulString, SideEffects> const& _functionSideEffects,
+		map<YulString, set<YulString>> const& _functionCalls
+	): m_dialect(_dialect), m_functionSideEffects(move(_functionSideEffects)), m_functionCalls(move(_functionCalls))
+	{}
+
+	using ASTModifier::operator();
+	void operator()(FunctionCall& _funCall) override
+	{
+		ASTModifier::operator()(_funCall);
+		if (m_seenCall)
+			util::BreadthFirstSearch<YulString>{{_funCall.functionName.name}, m_badFunctions}.run([&](YulString _fn, auto&& _addChild) {
```

Comment:
So yeah - this is the core part 1:
The side effect propagator already propagates invalidating returndatasize backwards through the call graph. This here now, propagates forwards, whenever the control flow I'm currently following has unknown returndatasize already.